packages/aws_common/lib/src/http/aws_http_client_io.dart

```diff
@@ -0,0 +1,522 @@
+      cancelTrigger.then((_) {
+        logger.verbose('Canceling request');
+        onCancel?.call();
+      }),
+    );
+
+    final redirects = <_RedirectInfo>[];
+    Future<AWSStreamedHttpResponse?> makeRequest(
+      AWSHttpMethod method,
+      Uri uri,
+    ) async {
+      final transport = _http2Connections[uri.authority] ??=
+          ClientTransportConnection.viaSocket(
+        await SecureSocket.connect(
+          uri.host,
+          uri.port,
+          supportedProtocols: supportedProtocols.alpnValues,
+          onBadCertificate: (cert) {
+            final certificate = X509Certificate(
+              der: cert.der,
+              pem: cert.pem,
+              sha1: cert.sha1,
+              subject: cert.subject,
+              issuer: cert.issuer,
+              startValidity: cert.startValidity,
+              endValidity: cert.endValidity,
+            );
+            return onBadCertificate(certificate, uri.host, uri.port);
+          },
+        ),
+      )..onActiveStateChanged = (isActive) {
+              if (!isActive) {
+                _logger.verbose('Closing transport: ${uri.authority}');
+                _http2Connections.remove(uri.authority)?.finish();
+              }
+            };
+      final stream = transport.makeRequest(
+        [
+          // Required pseudo-header fields as described here:
+          // https://www.rfc-editor.org/rfc/rfc7540#section-8.1.2.3
+          Header(':method'.codeUnits, utf8.encode(method.value)),
+          Header(
+            ':path'.codeUnits,
+            utf8.encode('${uri.path}${uri.hasQuery ? '?${uri.query}' : ''}'),
+          ),
+          Header(':scheme'.codeUnits, utf8.encode(uri.scheme)),
+          Header(
+            ':authority'.codeUnits,
+            utf8.encode('${uri.host}${uri.hasPort ? ':${uri.port}' : ''}'),
+          ),
+          for (final entry in request.headers.entries)
+            if (redirects.isEmpty ||
+                _shouldCopyHeaderOnRedirect(entry.key, request.uri, uri))
+              Header(utf8.encode(entry.key), utf8.encode(entry.value)),
+        ],
+      );
+
+      var requestBytesRead = 0;
+      request.body
+          .map((chunk) {
+            requestBytesRead += chunk.length;
+            requestProgress.add(requestBytesRead);
+            return DataStreamMessage(chunk);
+          })
+          .takeUntil(cancelTrigger)
+          .listen(
+            stream.outgoingMessages.add,
+            onDone: () {
+              logger.verbose('Request sent');
+              requestProgress.close();
+              stream.outgoingMessages.close();
+            },
+          );
+
+      final gotHeaders = Completer<void>();
+      onCancel = () {
+        if (!gotHeaders.isCompleted) {
+          gotHeaders.completeError(const CancellationException());
+        }
+      };
+      stream.onTerminated = (code) {
+        logger.debug('Stream terminated: $code');
+        _http2Connections.remove(uri.authority);
```

Comment:
question: Is this needed given above block should remove this when no active streams on the instance?

Reply:
Not needed - removed 👍